app/models/concerns/team_duplication.rb

```diff
@@ -7,215 +7,109 @@ module TeamDuplication
       end
     end
 
-    def self.copy_relationships(pm_mapping)
-      return if pm_mapping.blank?
-      Relationship.where(source_id: pm_mapping.keys).find_each do |r|
-        copy_r = r.dup
-        self.set_mapping(r, copy_r);
-        copy_r.is_being_copied = true
-        self.update_relationships(copy_r)
-        copy_r.save(validate: false)
-      end
-    end
-
-    def self.update_relationships(copy_r)
-      [:source_id, :target_id].each do |r|
-        pm_mapping = @mapping.dig(:ProjectMedia, copy_r.send(r))
-        copy_r.send("#{r}=", pm_mapping.id) if pm_mapping
-      end
-    end
-
-    def self.copy_versions(versions_mapping)
-      return if versions_mapping.blank?
-      versions_mapping.each_pair do |original, copy|
-        log = Version.find(original).dup
-        log.team_id = @copy_team.id
-        log.is_being_copied = true
-        log.associated_id = copy.id unless log.associated_id.blank?
-        item = @mapping.dig(log.item_type.to_sym, log.item_id.to_i)
-        self.update_version_fields(log, item)
-        log.save(validate: false)
+    def self.process_team_bot_installations(t, team)
+      processed_user_ids = []
+      t.team_bot_installations.each do |tbi|
+        new_tbi = tbi.deep_clone
+        new_tbi.team = team
+        if new_tbi.user.name == "Smooch"
+          new_tbi = Bot::Smooch.sanitize_installation(new_tbi)
+          new_tbi.settings["smooch_project_id"] = @project_id_map[tbi.settings["smooch_project_id"]]
+          new_tbi.settings["smooch_workflows"] = tbi.settings["smooch_workflows"]
+        end
+        new_tbi.save(validate: false)
+        processed_user_ids << new_tbi.user_id
       end
+      processed_user_ids
     end
 
-    def self.update_version_fields(log, item)
-      return unless item
-      self.update_version_object(log, item)
-      self.update_version_object_changes(log)
-      self.update_version_meta(log, item)
-      log.item_id = item.id
-      log.set_object_after
-    end
-
-    def self.update_version_object(log, item)
-      object = log.get_object
-      return if object.blank?
-      object['id'] = item.id if object['id']
-      object['annotated_id'] = item.annotated_id if object['annotated_id']
-      log.object = object.to_json
-    end
-
-    def self.update_version_object_changes(log)
-      changes = log.get_object_changes.with_indifferent_access
-      return if changes.blank?
-      associations = { annotated_id: 'associated_type', source_id: 'associated_type', target_id: 'associated_type', id: 'item_type' }
-      associations.each_pair do |field, method|
-        unless changes[field].blank?
-          changes[field].map! do |a|
-            c = @mapping.dig(log.send(method).to_sym, a)
-            c ? c.id : a
+    def self.store_clones(team)
+      @clones.each do |clone|
+        if !clone[:original].is_a?(Team)
+          if clone[:clone].respond_to?(:team_id) && clone[:clone].team_id.nil?
+            clone[:clone].team_id = team.id
```

Comment:
This line is missing a test.

Reply:
After more review I've determined that this is actually not a line we need - I don't actually think we should see a nil id here? I'm going to run a test with this out altogether.

Reply:
You can also replace the condition `&& clone[:clone].team_id.nil?` by a `clone[:clone].team_id ||= team.id` in the line below.